### PR TITLE
Support OpenSSL-based MsQuic on Windows

### LINKED ISSUE #69978: [QUIC] Support for OpenSSL build of MsQuic on Windows
Currently, we assume that on Windows, MsQuic always uses Schannel for TLS operations and on Linux it uses OpenSSL. This influences how we pass certificates to MsQuic

https://github.com/dotnet/runtime/blob/a4223dc25e6c0ff7da1185f0559b40d0ae75d3da/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs#L183-L221

And also some certificate related flags: https://github.com/dotnet/runtime/blob/a4223dc25e6c0ff7da1185f0559b40d0ae75d3da/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs#L55-L58

However, it is possible to build MsQuic with OpenSSL for Windows (e.g. to use on older Windows without TLS 1.3 support via Schannel). So we should instead make these decision based on the actual TLS backend being used. This information can be retrieved via recently added API: https://github.com/microsoft/msquic/pull/2762.

### PR DESCRIPTION
**Pull request by rzikm**
Closes #69978.

While not officially supported, it is not hard for us to support QUIC on older Windows version via OpenSSL-based build of MsQuic.

The only limitation now is that the `X509Certificate2` instances must have exportable private keys (i.e. passing `X509KeyStorageFlags.Exportable` to ctor).

### CODE REVIEW

**Comment by rzikm** [comment]
I tested the build with locally built MsQuic on Windows 11, it would be nice if somebody tested this e.g. on Windows 10 machine to confirm this.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs [hunk 1]
```diff
@@ -90,18 +76,39 @@ static MsQuicApi()
...
+                                    if (NetEventSource.Log.IsEnabled())
+                                    {
+                                        NetEventSource.Info(null, $"Current Windows version ({Environment.OSVersion}) is not supported by QUIC. Minimal supported version is {MinWindowsVersion}");
+                                    }
+
+                                    return;
+                                }
+
+                                Tls13ServerMayBeDisabled = IsTls13Disabled(true);
+                                Tls13ClientMayBeDisabled = IsTls13Disabled(false);
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs:107]:
```suggestion
                                Tls13ServerMayBeDisabled = IsTls13Disabled(isServer: true);
                                Tls13ClientMayBeDisabled = IsTls13Disabled(isServer: false);
```
I know you're just moving it and not changing, but it's always nice to have a name of parameter with the boolean constants.

**Review by ManickaP — COMMENTED** [review]
LGTM, but we should get a confirmation this works on Windows 10 before merging.

**Comment by wfurt** [comment]
> Does OpenSSL work well on windows?

AFAIK, MsQuic has tests for that scenario e.g. it is supported combination for them. Since this depend only on crypto, I see no reason why this would work any different than Linux. 
It will not work for .NET users out of the box but this allows anybody to bundle custom MsQuic bits.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs [hunk 2]
```diff
@@ -90,18 +76,39 @@ static MsQuicApi()
...
                                 {
                                     NetEventSource.Info(null, $"Incompatible MsQuic library version '{version}', expecting '{MsQuicVersion}'");
                                 }
+                                return;
+                            }
+
+                            // Assume SChanel is being used on windows and query for the actual provider from the library
+                            QUIC_TLS_PROVIDER provider = OperatingSystem.IsWindows() ? QUIC_TLS_PROVIDER.SCHANNEL : QUIC_TLS_PROVIDER.OPENSSL;
+                            size = sizeof(QUIC_TLS_PROVIDER);
+                            apiTable->GetParam(null, QUIC_PARAM_GLOBAL_TLS_PROVIDER, &size, &provider);
```
Comment (wfurt) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs:91]: Do we expect that the provider will not change if there is MsQuic version that does not support the query?

**Review by wfurt — APPROVED** [review]
LGTM

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs (hunk 2 above)
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs:91]: The question was really about running on older msquic versions. I don't know if this was added in/prior to 2.0 or if that is more recent addition.

Comment (nibanks) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs:91]: I think it's fine to assume this and state that MsQuic+OpenSSL is only supported on Windows for MsQuic v2+.

**Review by nibanks — APPROVED** [review]

**Comment by nibanks** [comment]
> Does OpenSSL work well on windows?

Yes, very well. It is is more feature complete than Schannel. The perf of OpenSSL crypto is not quite as good as in-box bcrypt though.

**Pull request merged**